### PR TITLE
Check the X-Forwarded-Proto header to determine if the status & file names should return https

### PR DESCRIPTION
**Pull request by lailasharshar**
### Summary
The status endpoint returns URLs (for both status checking and downloads) as http, not https. There are two places in BulkDataAccessAPI where we generate URLs that go back to the end user. In those places, we need to check if the Request contains X-Forwarded-Proto HTTP header set to https. If it does, replace http with https in the generated URLs.

Also, when we create Jobs we use ServletUriComponentsBuilder.fromCurrentRequest().toUriString()` which later comes back as part of the Status 200 body. We have to replace http with https as described above as well.

Updated BulkDataAccessAPIIntegrationTests to have a test case that passes that header and verifies https.

### Checklist

- [ ] Tests and linting pass
- [ ] Code checked for PHI/PII exposure

### Security
Adds https instead of http in places.

### CODE REVIEW

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 1]
```diff
@@ -189,15 +205,23 @@ private void logSuccessfulJobCreation(Job job) {
...
                 HttpStatus.ACCEPTED);
     }
 
+    String getUrl(String ending) {
+        String url = ServletUriComponentsBuilder.fromCurrentRequestUri().replacePath(ending).toUriString();
+        url = url.replace(":80/", "/");
+        if (!shouldReplaceWithHttps()) {
+            return url;
+        }
+        return url.replace("http://", "https://");
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: These lines duplicate those in `getCurrentUrl`; let's refactor into a common method.

File: api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java [hunk 2]
```diff
@@ -131,6 +131,30 @@ public void testBasicPatientExport() throws Exception {
         Assert.assertEquals(job.getUser(), userRepository.findByUsername(TEST_USER));
     }
 
+    @Test
+    public void testBasicPatientExportWithHttps() throws Exception {
+        ResultActions resultActions = this.mockMvc.perform(
+                get(API_PREFIX + FHIR_PREFIX + PATIENT_EXPORT_PATH).contentType(MediaType.APPLICATION_JSON)
+                        .header("Authorization", "Bearer " + token)
+                        .header("X-Forwarded-Proto", "https"))
+                .andDo(print());
```
Comment (dkrylovsb) [inline api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java]: Let's just make sure this does not print too much to console

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 3]
```diff
@@ -124,14 +125,29 @@
...
+        Job job = jobService.createJob(resourceTypes, getCurrentUrl(), null, outputFormat, since);
 
         logSuccessfulJobCreation(job);
 
         return returnStatusForJobCreation(job);
     }
 
+    private String getCurrentUrl() {
+        String request = ServletUriComponentsBuilder.fromCurrentRequest().toUriString();
+        request = request.replace(":80/", "/");
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: I did not see `80` anywhere within the URLs, in which case we _might_ not need this replacement

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 4]
```diff
@@ -124,14 +125,29 @@
...
+        if (!shouldReplaceWithHttps()) {
+            return request;
+        }
+        request = request.replace("http://", "https://");
```
Comment (dkrylovsb) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: If you'd like to do this replacement in a more OOP way, `org.springframework.web.util.UriComponentsBuilder#scheme` method will do the trick

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java (hunk 3 above)
Comment (lailasharshar) [reply api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: Oh we do. For some reason, at some point, it ends up creating the url https://localhost:80.

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java (hunk 4 above)
Comment (lailasharshar) [reply api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: Ah, thanks. That is what I was looking for. I just spent so much time googling for it unsuccessfully, that I figured I would put it out there. Maybe that will solve it coming back with :80/

File: api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java (hunk 2 above)
Comment (lailasharshar) [reply api/src/test/java/gov/cms/ab2d/api/controller/BulkDataAccessAPIIntegrationTests.java]: Fair enough, but it's been there forever and I actually found it really helpful.

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java (hunk 3 above)
Comment (lailasharshar) [reply api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: And adding scheme doesn't seem to fix that. It's when it's generating the file names but the request comes in as http.

File: api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java [hunk 5]
```diff
@@ -124,14 +125,25 @@
         checkResourceTypesAndOutputFormat(resourceTypes, outputFormat);
         checkSinceTime(since);
 
-        Job job = jobService.createJob(resourceTypes, ServletUriComponentsBuilder.fromCurrentRequest().toUriString(),
-                null, outputFormat, since);
+        Job job = jobService.createJob(resourceTypes, getCurrentUrl(), null, outputFormat, since);
```
Comment (spathiyil) [inline api/src/main/java/gov/cms/ab2d/api/controller/BulkDataAccessAPI.java]: nice clean up here. I am able to read it better.

**Pull request merged**